File: apps/web/src/types/progress.ts
export interface TopicProgress {
  id: string;
  name: string;
  category: string;
  attempts: number;
  correct_attempts: number;
  average_score: number;
  mastery_score: number;
  last_score: number;
  last_seen_at: string;
}

export interface KnowledgeEdge {
  id: string;
  source_topic_id: string;
  target_topic_id: string;
  edge_type: string;
  weight: number;
}

export interface LearnerProgress {
  learner_id: string;
  sessions_completed: number;
  answers_reviewed: number;
  average_score: number;
  recommended_difficulty: "easy" | "medium" | "hard" | string;
  topics: TopicProgress[];
  edges: KnowledgeEdge[];
}

File: apps/api/app/schemas/progress.py
"""Learner progress and knowledge graph response contracts."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from app.schemas.common import AptlyBaseModel


class TopicProgressResponse(AptlyBaseModel):
    id: UUID
    name: str
    category: str
    attempts: int
    correct_attempts: int
    average_score: float
    mastery_score: float
    last_score: float
    last_seen_at: datetime


class KnowledgeEdgeResponse(AptlyBaseModel):
    id: UUID
    source_topic_id: UUID
    target_topic_id: UUID
    edge_type: str
    weight: int


class ProgressResponse(AptlyBaseModel):
    learner_id: str
    sessions_completed: int
    answers_reviewed: int
    average_score: float
    recommended_difficulty: str
    topics: list[TopicProgressResponse]
    edges: list[KnowledgeEdgeResponse]
